File: solutions/python3/0082.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy = ListNode(None)
        prev = dummy
        dummy.next = head

        while head:
            while head.next and head.val == head.next.val:
                head = head.next
            if prev.next == head:
                prev = prev.next
            else:
                prev.next = head.next
            head = head.next

        return dummy.next

File: solutions/cpp/0082.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* dummy = new ListNode(NULL);
        ListNode* prev = dummy;
        dummy->next = head;

        while (head) {
            while (head->next && head->val == head->next->val)
                head = head->next;
            if (prev->next == head)
                prev = prev->next;
            else
                prev->next = head->next;
            head = head->next;
        }

        return dummy->next;
    }
};